compiler suggests invalid code with ref pattern

The following example:

```rust
struct S {
    f1: i32,
}

fn main() {
    let s = S { f1: 123 };
    let S { ref f1 } = s;
}
```

Gives the following suggestion:

```
warning: unused variable: `f1`
 --> src/main.rs:9:15
  |
9 |     let S{ref f1} = s;
  |               ^^ help: try ignoring the field: `f1: _`
  |
  = note: #[warn(unused_variables)] on by default
```

When the suggestion is applied, the code does not compile:

```
error: expected `,`
 --> src/main.rs:9:17
  |
9 |     let S { ref f1: _ } = s;
  |                 ^^

error[E0027]: pattern does not mention field `f1`
 --> src/main.rs:9:9
  |
9 |     let S { ref f1: _ } = s;
  |         ^^^^^^^^^^^^^^^ missing field `f1`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0027`.
```

I'm not sure what the best suggestion would be.  Removing the `ref` might be an option (I'm not sure if using a wildcard is equivalent in this case?).  Suggesting `..` is another option.

This suggestion was changed in 1.28 (previous versions gave a different, wrong suggestion).  Tested up to rustc 1.30.0-nightly (2d4e34ca8 2018-09-09).

--- a/src/librustc/middle/liveness.rs
+++ b/src/librustc/middle/liveness.rs
@@ -1627,11 +1627,18 @@ impl<'a, 'tcx> Liveness<'a, 'tcx> {
                     );
 
                     if self.ir.variable_is_shorthand(var) {
-                        err.multipart_suggestion(
-                            "try ignoring the field",
-                            spans.iter().map(|span| (*span, format!("{}: _", name))).collect(),
-                            Applicability::MachineApplicable
-                        );
+                        if let Node::Binding(pat) = self.ir.tcx.hir().get_by_hir_id(hir_id) {
+                            // Handle `ref` and `ref mut`.
+                            let spans = spans.iter()
+                                .map(|_span| (pat.span, format!("{}: _", name)))
+                                .collect();
+
+                            err.multipart_suggestion(
+                                "try ignoring the field",
+                                spans,
+                                Applicability::MachineApplicable,
+                            );
+                        }
                     } else {
                         err.multipart_suggestion(
                             "consider prefixing with an underscore",

--- a/src/test/ui/lint/issue-54180-unused-ref-field.fixed
+++ b/src/test/ui/lint/issue-54180-unused-ref-field.fixed
@@ -1,0 +1,34 @@
+// run-rustfix
+
+#![deny(unused)]
+
+pub struct S {
+    pub f1: i32,
+}
+
+pub struct Point {
+    pub x: i32,
+    pub y: i32,
+}
+
+pub enum E {
+    Variant { field: String }
+}
+
+pub fn foo(arg: &E) {
+    match arg {
+        E::Variant { field: _ } => (), //~ ERROR unused variable
+    }
+}
+
+fn main() {
+    let s = S { f1: 123 };
+    let S { f1: _ } = s; //~ ERROR unused variable
+
+    let points = vec![Point { x: 1, y: 2 }];
+    let _: i32 = points.iter().map(|Point { x: _, y }| y).sum(); //~ ERROR unused variable
+
+    match (Point { x: 1, y: 2 }) {
+        Point { y, x: _ } => y, //~ ERROR unused variable
+    };
+}

--- a/src/test/ui/lint/issue-54180-unused-ref-field.rs
+++ b/src/test/ui/lint/issue-54180-unused-ref-field.rs
@@ -1,0 +1,34 @@
+// run-rustfix
+
+#![deny(unused)]
+
+pub struct S {
+    pub f1: i32,
+}
+
+pub struct Point {
+    pub x: i32,
+    pub y: i32,
+}
+
+pub enum E {
+    Variant { field: String }
+}
+
+pub fn foo(arg: &E) {
+    match arg {
+        E::Variant { ref field } => (), //~ ERROR unused variable
+    }
+}
+
+fn main() {
+    let s = S { f1: 123 };
+    let S { ref f1 } = s; //~ ERROR unused variable
+
+    let points = vec![Point { x: 1, y: 2 }];
+    let _: i32 = points.iter().map(|Point { x, y }| y).sum(); //~ ERROR unused variable
+
+    match (Point { x: 1, y: 2 }) {
+        Point { y, ref mut x } => y, //~ ERROR unused variable
+    };
+}

--- a/src/test/ui/lint/issue-54180-unused-ref-field.stderr
+++ b/src/test/ui/lint/issue-54180-unused-ref-field.stderr
@@ -1,0 +1,39 @@
+error: unused variable: `field`
+  --> $DIR/issue-54180-unused-ref-field.rs:20:26
+   |
+LL |         E::Variant { ref field } => (),
+   |                      ----^^^^^
+   |                      |
+   |                      help: try ignoring the field: `field: _`
+   |
+note: lint level defined here
+  --> $DIR/issue-54180-unused-ref-field.rs:3:9
+   |
+LL | #![deny(unused)]
+   |         ^^^^^^
+   = note: #[deny(unused_variables)] implied by #[deny(unused)]
+
+error: unused variable: `x`
+  --> $DIR/issue-54180-unused-ref-field.rs:29:45
+   |
+LL |     let _: i32 = points.iter().map(|Point { x, y }| y).sum();
+   |                                             ^ help: try ignoring the field: `x: _`
+
+error: unused variable: `f1`
+  --> $DIR/issue-54180-unused-ref-field.rs:26:17
+   |
+LL |     let S { ref f1 } = s;
+   |             ----^^
+   |             |
+   |             help: try ignoring the field: `f1: _`
+
+error: unused variable: `x`
+  --> $DIR/issue-54180-unused-ref-field.rs:32:28
+   |
+LL |         Point { y, ref mut x } => y,
+   |                    --------^
+   |                    |
+   |                    help: try ignoring the field: `x: _`
+
+error: aborting due to 4 previous errors
+

